channel/event.go

```diff
@@ -0,0 +1,183 @@
+package channel
+
+import (
+	"log"
+	"slices"
+	"time"
+
+	"github.com/dop251/goja"
+)
+
+type event struct {
+	id         int32
+	duration   time.Duration
+	endTime    time.Time
+	finished   chan struct{}
+	instanceID int
+	playerIDs  []int32
+	server     *Server
+
+	startCallback            func()
+	beforePortalCallback     func(plr scriptPlayerWrapper, src scriptMapWrapper, dst scriptMapWrapper) bool
+	afterPortalCallback      func(plr scriptPlayerWrapper, dst scriptMapWrapper)
+	timeoutCallback          func(plr scriptPlayerWrapper)
+	playerLeaveEventCallback func(plr scriptPlayerWrapper)
+
+	program *goja.Program
+	vm      *goja.Runtime
+}
+
+func createEvent(id int32, instID int, players []int32, server *Server, program *goja.Program) *event {
+	ctrl := &event{
+		id:         id,
+		finished:   make(chan struct{}),
+		instanceID: instID,
+		playerIDs:  players,
+		server:     server,
+		program:    program,
+		vm:         goja.New(),
+	}
+
+	ctrl.vm.SetFieldNameMapper(goja.UncapFieldNameMapper())
+	_ = ctrl.vm.Set("ctrl", ctrl)
+
+	_, err := ctrl.vm.RunProgram(ctrl.program)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	err = ctrl.vm.ExportTo(ctrl.vm.Get("start"), &ctrl.startCallback)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	err = ctrl.vm.ExportTo(ctrl.vm.Get("beforePortal"), &ctrl.beforePortalCallback)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	err = ctrl.vm.ExportTo(ctrl.vm.Get("afterPortal"), &ctrl.afterPortalCallback)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	err = ctrl.vm.ExportTo(ctrl.vm.Get("playerLeaveEvent"), &ctrl.playerLeaveEventCallback)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	err = ctrl.vm.ExportTo(ctrl.vm.Get("timeout"), &ctrl.timeoutCallback)
+
+	if err != nil {
+		log.Println(err)
+	}
+
+	return ctrl
+}
+
+func (e *event) start(server *Server) {
+	e.startCallback()
+
+	for _, id := range e.playerIDs {
+		if plr, err := e.server.players.GetFromID(id); err == nil {
+			plr.event = e
+		}
+	}
+
+	go func() {
+		timeout := time.NewTimer(e.duration)
+		defer timeout.Stop()
+
+		select {
+		case <-timeout.C:
+			server.dispatch <- func() {
+				for _, id := range e.playerIDs {
+					if plr, err := server.players.GetFromID(id); err == nil {
+						plr.event = nil
+						e.timeoutCallback(scriptPlayerWrapper{plr: plr, server: e.server})
+					}
+				}
+
+				delete(server.events, e.id)
+			}
+		case <-e.finished:
+			server.dispatch <- func() {
+				for _, id := range e.playerIDs {
+					if plr, err := server.players.GetFromID(id); err == nil {
+						plr.event = nil
+					}
+				}
+
+				delete(server.events, e.id)
+			}
+			return
+		}
+	}()
+}
+
+func (e *event) Log(msg string) {
+	log.Println(msg)
+}
+
+func (e *event) RemainingTime() int32 {
+	return int32(time.Until(e.endTime).Seconds())
+}
+
+func (e *event) PlayerCount() int {
+	return len(e.playerIDs)
+}
+
+func (e *event) Finished() {
+	close(e.finished)
```

Comment:
Should probably use sync.Once here to prevent possible panics.

Reply:
Good catch